### ISSUE TITLE
going offline after some time

### ISSUE THREAD

**Issue by Jaglag** [markstory/hubot-xmpp#63]
Hi,

the bot just going offline after some time of idle and i dont know how to fix that:

```
[Sat Feb 22 2014 13:41:31 GMT+0100 (CET)] DEBUG [received iq]  iq type="get" id="111-111111" from="robertsspaceindustries.com" to="usernam@robertsspaceindustries.com/7fbb6dc0" xmlns:stream="http://etherx.jabber.org/streams"><ping xmlns="urn:xmpp:ping"/ /iq


 [Sat Feb 22 2014 13:41:31 GMT+0100 (CET)] DEBUG [sending pong] iq to="robertsspaceindustries.com" from="usernam@robertsspaceindustries.com/7fbb6dc0" type="result" id="111-111111"/


 [Sat Feb 22 2014 13:46:34 GMT+0100 (CET)] DEBUG Received offline event
 [Sat Feb 22 2014 13:46:34 GMT+0100 (CET)] DEBUG Received offline event
```

sometimes it takes 30min sometime over 2 hours then he´s going offline

**Comment by pixelrebel** [comment]
I'm having this issue too.  But I don't see the server sending an offline event.  Instead, what I see in the log is that hubot sends a ping every 30 seconds and receives a bad-request "Missing 'id'" error:
```
[Wed Aug 26 2015 13:10:54 GMT-0700 (PDT)] DEBUG [sending ping] <iq type="get"><ping xmlns="urn:xmpp:ping"/></iq>
[Wed Aug 26 2015 13:10:54 GMT-0700 (PDT)] ERROR [xmpp error]<iq type="error" xmlns:stream="http://etherx.jabber.org/streams"><error type="modify"><bad-request xmlns="urn:ietf:params:xml:ns:xmpp-stanzas"/><text xmlns="urn:ietf:params:xml:ns:xmpp-stanzas" xml:lang="en">Missing 'id'</text></error></iq>
```
When hubot disconnects nothing else shows in the logs, except it failed to send a ping.  It's the only 60 second gap in the log.  Then it continues to fruitlessly ping every 30 seconds and receiving the bad-request error.

**Comment by pixelrebel** [comment]
Still stuck on the above issue.  Can anyone confirm this "Missing id" error is causing hubot to disconnect? My workaround is a cron restart every 30 minutes.  Any advice?  Thanks.

**Comment by anupdhml** [comment]
@pixelrebel are you using this with slack? There's a slack adapter for hubot, which should be a better choice: https://github.com/slackhq/hubot-slack 

If you have to use this, your problem seems to be beacuse hubot-xmpp currently does not include 'id' field in the ping messages: https://github.com/markstory/hubot-xmpp/blob/master/src/xmpp.coffee#L112

**Comment by pixelrebel** [comment]
@anupdhml: I am using the hubot-xmpp adapter with slack because my bot needs to upload files and send attachments which (correct me if I'm wrong) isn't currently supported by hubot-slack.

Perhaps I can fork this project and add a line for the id.  Do you know what id this error is referring to?

**Comment by pixelrebel** [comment]
Well, adding the id to iq ping stanza didn't solve my disconnection problem, but it does make the "bad-request" error go away.  

I submitted a PR for this fix:
https://github.com/markstory/hubot-xmpp/pull/93

**Comment by anupdhml** [comment]
@pixelrebel I found this for slack: https://github.com/slackhq/node-slack-client/pull/56, but merging it could be tricky. hubot-xmpp does not currently support sending attachments though: you would have to write in the functionality. So maybe give hubot-slack with that PR a try?

From the [xmpp specification](http://xmpp.org/rfcs/rfc3920.html#streams):
> There SHOULD NOT be an 'id' attribute on the XML stream header sent from the initiating entity to the receiving entity; however, if an 'id' attribute is included, it SHOULD be silently ignored by the receiving entity. 

But looks like slack has different expectations. Adding id to ping stanzas probably won't have any side effects, so your PR should be all right. I used ids in https://github.com/markstory/hubot-xmpp/pull/90#issuecomment-143859232 and haven't noticed anything off.

**Comment by pixelrebel** [comment]
@anupdhml That's interesting about the xmpp spec.  I actually found [this example](http://xmpp.org/extensions/xep-0099.html) which includes the `id in the iq tag, but I guess it's not based off the official spec?

Now I'm thinking it's best abide by the spec.  I can just keep my own fork if I continue to use this adapter with slack.  In fact, https://github.com/markstory/hubot-xmpp/pull/93 does not fix my connection problem, so it just a petty fix to avoid an error from slack.

**Comment by anupdhml** [comment]
@pixelrebel yes https://github.com/markstory/hubot-xmpp/pull/93 does not solve your connection issues. Have you tried debugging from here: https://github.com/markstory/hubot-xmpp/blob/master/src/xmpp.coffee#L194

Add something like `console.log stanza.toString()` to see all the incoming stanzas from the server. Maybe it will give you a clue as to what's going wrong. If only you could access slack logs, this would be much easier.

**Comment by pixelrebel** [comment]
Thanks for the tip @anupdhml .  I gave that a try, but I didn't see anything that stood out.  I have a support ticket open asking them to check their server logs around the last time I was booted.  I'll let you know if they come back with anything interesting.

I'm thinking about going back to the [hubot-slack](https://github.com/slackhq/hubot-slack) adapter.  This recent commit https://github.com/slackhq/hubot-slack/commit/d8d6660a7cbfa55f90800a75e0e1b9367ccfc8b4 is sort of a revelation to me.  It seems slack now allows real users to use the adapter.  Which is all I really need at this point.  Attachments are also supported, and I'm using curl to upload files anyway.

**Comment by pixelrebel** [comment]
@anupdhml Here's what slack support came back with:
>It looks like the client is disconnecting — there was a pretty noticeable pattern of disconnection messages coming in every thirty minutes. It also looks like a lot of these disconnections are initiated around the same time the client is marked as away. I'm not sure that's related, but thought I'd pass it on anyway.
This makes sense to me because it doesn't seem to disconnect when the bot is actively posting.  So, is there a way to keep the bot from going idle?

**Comment by anupdhml** [comment]
@pixelrebel If that's the case, you should be able to keep hubot online by broadcasting presence messages at fixed intervals.

http://xmpp.org/rfcs/rfc3921.html#rfc.section.5.1.2 

If slack's XMPP implementation follows the spec here, it should work. You can model the code on the [initial presence stanza](https://github.com/markstory/hubot-xmpp/blob/master/src/xmpp.coffee#L100 ) that hubot currently sends.

There should be a better way of doing it though: the underlying [node-xmpp client](https://github.com/node-xmpp/node-xmpp-client/blob/master/lib/Client.js) should have detected the disconnect from slack, and hubot-xmpp should be able to catch that and reconnect. There's already some code for [reconnecting](https://github.com/markstory/hubot-xmpp/blob/master/src/xmpp.coffee#L85), but looks like slack behaves differently. See if the disconnect event used [here](https://github.com/node-xmpp/node-xmpp-client/blob/6e0ad6f7790d997e16b353e987a9f2528044c2e6/examples/question_bot.js#L72) helps.

Let me know if you need help as you go about this, good luck!
